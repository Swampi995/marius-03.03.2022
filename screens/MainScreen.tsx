import { useState } from 'react';
import { StyleSheet } from 'react-native';
import { View } from '../components';
import Header from './book/Header';
import Actions from './book/Actions';
import Spread from './book/Spread';
import Asks from './book/Asks';
import Bids from './book/Bids';
import { provideMarketData, MARKET, OrdersData } from '../services/ordersData';

export interface Orders {
  [key: number]: {
    price: number,
    size: number,
    total: number
  }
}

export default function MainScreen() {
  const [group, setGroup] = useState(0.5);
  const [killed, setKilled] = useState(false);
  const [market, setMarket] = useState<MARKET>('PI_XBTUSD');
  const marketHook = provideMarketData(market);

  const toggleMaket = (newMarket: MARKET) => {
    setMarket(newMarket);
    if (newMarket === 'PI_XBTUSD') {
      setGroup(0.5);
    } else {
      setGroup(0.05);
    }

    marketHook.subscribe(newMarket);
  }

  const toggleKilled = (newKilled: boolean) => {
    setKilled(newKilled);

    if (!newKilled) {
      marketHook.stop();
    } else {
      marketHook.start();
    }
  }

  const { items: asks, maxTotal: askMax, min: minAsk } = processOrders(marketHook.data?.asks, group);
  const { items: bids, maxTotal: bidMax, max: maxBid } = processOrders(marketHook.data?.bids, group);

  const maxTotal = Math.max(askMax, bidMax);

  return (
    <View style={styles.container}>
      <Header market={market} group={group} changeGroup={setGroup} />
      <Asks asks={asks} maxTotal={maxTotal} />
      <Spread maxBid={maxBid} minAsk={minAsk} />
      <Bids bids={bids} maxTotal={maxTotal} />
      <Actions market={market} setMarket={toggleMaket} killed={killed} setKilled={toggleKilled} />
    </View>
  );
}

export const processOrders = (orders: OrdersData['asks'] | OrdersData['bids'] | undefined, group: number) => {
  if (!orders?.length) {
    return { items: {}, maxTotal: 0, max: 0, min: 0 };
  }

  const items = orders.reduce((all: Orders, [price, size]) => {
    if (size !== 0) {
      const matchGroup = Object.keys(all).find((p) => Math.abs(parseFloat(p) - price) <= group);
      if (!matchGroup) {
        all[price] = {
          price,
          size,
          total: 0,
        }
      } else {
        all[parseFloat(matchGroup)].size += size;
      }
    }

    return all;
  }, {}) || {};

  let max = 0;
  let min = Number.MAX_SAFE_INTEGER;
  let maxTotal = 0;
  const sortedKeys = Object.keys(items).sort((a, b) => parseFloat(a) - parseFloat(b));
  sortedKeys.map((key, index) => {
    const lastKey = parseFloat(sortedKeys[index - 1]);
    items[parseFloat(key)].total = items[parseFloat(key)].size + (items[lastKey]?.total || 0);
    maxTotal = Math.max(items[parseFloat(key)].total, maxTotal);
    max = Math.max(items[parseFloat(key)].price, max);
    min = Math.min(items[parseFloat(key)].price, min);
  });

  return { items, maxTotal, max, min };
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});
