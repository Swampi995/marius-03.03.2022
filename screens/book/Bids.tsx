import { FC } from 'react';
import { StyleSheet } from 'react-native';
import { Text, View, useThemeColor } from '../../components';
import { Orders } from '../MainScreen';

interface BidsProps {
    maxTotal: number;
    bids: Orders;
}

const Bids: FC<BidsProps> = ({ maxTotal, bids }) => {
    const bidColor = useThemeColor('bid');
    const greenColor = useThemeColor('green');
    const whiteColor = useThemeColor('white');
    const limitedKeys = Object.keys(bids).sort((a, b) => parseFloat(a) - parseFloat(b)).slice(-12);

    return (
        <View style={[styles.container]}>
            {limitedKeys.map((key) => {
                const bid = bids[parseFloat(key)];
                const width = (bid.total * 100) / maxTotal;
                return <View key={bid.price} style={[styles.item]}>
                    <View style={[styles.bid, { backgroundColor: bidColor, width: `${width}%` }]}></View>
                    <Text style={[styles.text, { flex: 3, color: greenColor }]}>{bid.price.toLocaleString("en-US")}</Text>
                    <Text style={[styles.text, { color: whiteColor }]}>{bid.size.toLocaleString("en-US")}</Text>
                    <Text style={[styles.text, { color: whiteColor }]}>{bid.total.toLocaleString("en-US")}</Text>
                </View>
            })}
        </View>
    );
}

export default Bids;

const styles = StyleSheet.create({
    container: {
        flex: 8,
        overflow: 'hidden',
    },
    text: {
        paddingVertical: 4,
        flex: 2,
        textAlign: 'center'
    },
    item: {
        flexDirection: 'row',
    },
    bid: {
        height: '100%',
        position: 'absolute',
    }
});
