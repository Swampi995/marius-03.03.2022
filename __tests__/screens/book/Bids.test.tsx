import renderer from 'react-test-renderer';

import Bids from '../../../screens/book/Bids';
import { Orders } from '../../../screens/MainScreen';

describe('<Bids />', () => {
    const mockOrders: Orders = {
        10: {
            price: 10,
            size: 100,
            total: 100,
        },
        150: {
            price: 150,
            size: 300,
            total: 400,
        },
        500: {
            price: 500,
            size: 500,
            total: 1000,
        },
    };

    it('renders correctly', () => {
        const tree = renderer.create(<Bids maxTotal={1000} bids={mockOrders} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});