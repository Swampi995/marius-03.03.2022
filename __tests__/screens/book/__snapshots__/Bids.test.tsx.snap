// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Bids /> renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#121723",
      },
      Array [
        Object {
          "flex": 8,
          "overflow": "hidden",
        },
      ],
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#121723",
        },
        Array [
          Object {
            "flexDirection": "row",
          },
        ],
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#121723",
          },
          Array [
            Object {
              "height": "100%",
              "position": "absolute",
            },
            Object {
              "backgroundColor": "#11835d7d",
              "width": "10%",
            },
          ],
        ]
      }
    />
    <Text
      style={
        Array [
          Object {
            "color": "#bebfc5",
            "fontFamily": undefined,
          },
          Array [
            Object {
              "flex": 2,
              "paddingVertical": 4,
              "textAlign": "center",
            },
            Object {
              "color": "#11835d",
              "flex": 3,
            },
          ],
        ]
      }
    >
      10
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#bebfc5",
            "fontFamily": undefined,
          },
          Array [
            Object {
              "flex": 2,
              "paddingVertical": 4,
              "textAlign": "center",
            },
            Object {
              "color": "#cbced4",
            },
          ],
        ]
      }
    >
      100
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#bebfc5",
            "fontFamily": undefined,
          },
          Array [
            Object {
              "flex": 2,
              "paddingVertical": 4,
              "textAlign": "center",
            },
            Object {
              "color": "#cbced4",
            },
          ],
        ]
      }
    >
      100
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#121723",
        },
        Array [
          Object {
            "flexDirection": "row",
          },
        ],
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#121723",
          },
          Array [
            Object {
              "height": "100%",
              "position": "absolute",
            },
            Object {
              "backgroundColor": "#11835d7d",
              "width": "40%",
            },
          ],
        ]
      }
    />
    <Text
      style={
        Array [
          Object {
            "color": "#bebfc5",
            "fontFamily": undefined,
          },
          Array [
            Object {
              "flex": 2,
              "paddingVertical": 4,
              "textAlign": "center",
            },
            Object {
              "color": "#11835d",
              "flex": 3,
            },
          ],
        ]
      }
    >
      150
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#bebfc5",
            "fontFamily": undefined,
          },
          Array [
            Object {
              "flex": 2,
              "paddingVertical": 4,
              "textAlign": "center",
            },
            Object {
              "color": "#cbced4",
            },
          ],
        ]
      }
    >
      300
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#bebfc5",
            "fontFamily": undefined,
          },
          Array [
            Object {
              "flex": 2,
              "paddingVertical": 4,
              "textAlign": "center",
            },
            Object {
              "color": "#cbced4",
            },
          ],
        ]
      }
    >
      400
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#121723",
        },
        Array [
          Object {
            "flexDirection": "row",
          },
        ],
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#121723",
          },
          Array [
            Object {
              "height": "100%",
              "position": "absolute",
            },
            Object {
              "backgroundColor": "#11835d7d",
              "width": "100%",
            },
          ],
        ]
      }
    />
    <Text
      style={
        Array [
          Object {
            "color": "#bebfc5",
            "fontFamily": undefined,
          },
          Array [
            Object {
              "flex": 2,
              "paddingVertical": 4,
              "textAlign": "center",
            },
            Object {
              "color": "#11835d",
              "flex": 3,
            },
          ],
        ]
      }
    >
      500
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#bebfc5",
            "fontFamily": undefined,
          },
          Array [
            Object {
              "flex": 2,
              "paddingVertical": 4,
              "textAlign": "center",
            },
            Object {
              "color": "#cbced4",
            },
          ],
        ]
      }
    >
      500
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#bebfc5",
            "fontFamily": undefined,
          },
          Array [
            Object {
              "flex": 2,
              "paddingVertical": 4,
              "textAlign": "center",
            },
            Object {
              "color": "#cbced4",
            },
          ],
        ]
      }
    >
      1,000
    </Text>
  </View>
</View>
`;
